activerecord/test/cases/database_statements_test.rb

```diff
@@ -22,6 +22,37 @@ def test_create_should_return_the_inserted_id
     assert_not_nil return_the_inserted_id(method: :create)
   end
 
+  if current_adapter?(:PostgreSQLAdapter)
+    def test_exec_insert_returns_result_with_values_for_returning_columns
```

Comment:
Is `exec_insert` independently test-worthy? I note it's `nodoc`, so shouldn't be part of public API.

Totally open to keeping this if you think it helps keep the implementation in line, etc... just wanted to note that it's probably optional from a "we should test our exposed API behaviour" perspective. (Unless we already have notable coverage, in which case it's probably better to follow that precedent now, and ask any test-worthiness questions more holistically, and later.)

Reply:
It's only Trilogy's one that is `:nodoc:` but the "parent" (abstract) one is documented so I believe it fair to expect applications use it despite of it being really low-level compared to other ways of performing an insert
https://github.com/rails/rails/blob/3d84e974916412bb4ed7625fa99aacdb2bc1c71e/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb#L145-L151

https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/DatabaseStatements.html#method-i-exec_insert

Generally I do avoid testing private APIs in favor of testing behavior through public APIs but in this case it does look like `exec_insert` is a part of the public API